ush/rocoto/setup_expt.py

```diff
@@ -79,218 +112,162 @@ def edit_baseconfig():
+        "@MODE@": inputs.mode,
+        "@CHGRP_RSTPROD@": host.info["chgrp_rstprod"],
+        "@CHGRP_CMD@": host.info["chgrp_cmd"],
+        "@HPSSARCH@": host.info["hpssarch"],
+        "@gfs_cyc@": inputs.gfs_cyc,
+        "@APP@": inputs.app,
+    }
+
+    if inputs.mode in ['cycled']:
+        extend_dict = {
+            "@CASEENS@": f'C{inputs.resens}',
+            "@NMEM_ENKF@": inputs.nens,
+        }
+        tmpl_dict = dict(tmpl_dict, **extend_dict)
+
     with open(base_config + '.emc.dyn', 'rt') as fi:
-        with open(base_config, 'wt') as fo:
-            for line in fi:
-                line = line.replace('@MACHINE@', machine.upper()) \
-                    .replace('@PSLOT@', pslot) \
-                    .replace('@SDATE@', idate.strftime('%Y%m%d%H')) \
-                    .replace('@FDATE@', fdate.strftime('%Y%m%d%H')) \
-                    .replace('@EDATE@', edate.strftime('%Y%m%d%H')) \
-                    .replace('@CASEENS@', f'C{resens}') \
-                    .replace('@CASECTL@', f'C{resdet}') \
-                    .replace('@NMEM_ENKF@', f'{nens}') \
-                    .replace('@HOMEgfs@', top) \
-                    .replace('@BASE_GIT@', base_git) \
-                    .replace('@DMPDIR@', dmpdir) \
-                    .replace('@NWPROD@', nwprod) \
-                    .replace('@COMROOT@', comroot) \
-                    .replace('@HOMEDIR@', homedir) \
-                    .replace('@STMP@', stmp) \
-                    .replace('@PTMP@', ptmp) \
-                    .replace('@NOSCRUB@', noscrub) \
-                    .replace('@ACCOUNT@', account) \
-                    .replace('@QUEUE@', queue) \
-                    .replace('@QUEUE_SERVICE@', queue_service) \
-                    .replace('@PARTITION_BATCH@', partition_batch) \
-                    .replace('@EXP_WARM_START@', exp_warm_start) \
-                    .replace('@MODE@', 'cycled') \
-                    .replace('@CHGRP_RSTPROD@', chgrp_rstprod) \
-                    .replace('@CHGRP_CMD@', chgrp_cmd) \
-                    .replace('@HPSSARCH@', hpssarch) \
-                    .replace('@gfs_cyc@', f'{gfs_cyc}') \
-                    .replace('@APP@', f'{app}')
-                if expdir is not None:
-                    line = line.replace('@EXPDIR@', os.path.dirname(expdir))
-                if comrot is not None:
-                    line = line.replace('@ROTDIR@', os.path.dirname(comrot))
-                if 'ICSDIR' in line:
-                    continue
-                fo.write(line)
+        basestr = fi.read()
+
+    for key, val in tmpl_dict.items():
+        basestr = basestr.replace(key, str(val))
+
+    with open(base_config, 'wt') as fo:
+        fo.write(basestr)
 
     print('')
-    print(f'EDITED:  {expdir}/config.base as per user input.')
-    print(f'DEFAULT: {expdir}/config.base.emc.dyn is for reference only.')
+    print(f'EDITED:  {inputs.expdir}/config.base as per user input.')
```

Comment:
A minor point.  This should remain `{expdir}` as `{inputs.expdir}` points to `$EXPDIR/..`.

Reply:
I don't think expdir is defined in this namespace. {base_config} should be used anyway.

Reply:
`expdir` is defined in the namespace since it is part of `__main__`.  `base_config` works for line 163, but not 164 (apologies for not highlighting 164 in the above snippet).  However, a similar assignment could be used for config.base.emc.dyn.

Reply:
Ah, that works.

Reply:
Should be fixed now.